base/compiler/typeinfer.jl

```diff
@@ -320,12 +320,18 @@ function CodeInstance(interp::AbstractInterpreter, result::InferenceResult,
         end
     end
     relocatability = 0x0
-    if isa(inferred_result, String)
-        t = @_gc_preserve_begin inferred_result
-        relocatability = unsafe_load(unsafe_convert(Ptr{UInt8}, inferred_result), Core.sizeof(inferred_result))
-        @_gc_preserve_end t
-    elseif inferred_result === nothing
+    if const_flags == 0x3
```

Comment:
```suggestion
    if const_flags == 0x3 && may_discard_trees(interp)
```
Some external `AbstractInterpreter`s, such as Cthulhu and JET, want to always store additional information within `CodeInstance`s owned by them. Currently, they achieve this by using `transform_result_for_cache` to convert `InferenceResult` into their own specialized caching data structure. While it may be possible for them to recover this information when needed (similar to what we're doing with `codeinfo_for_const` in this PR), a more straightforward approach might be to keep `may_discard_trees(interp) === false` to ensure that `transform_result_for_cache` is always called.

Reply:
SGTM, but we may need to tweak this again in the future, as it will result in Cthulhu's display not matching `@code_typed`.